lib/jpegstream.js

```diff
@@ -30,33 +31,43 @@ var Stream = require('stream').Stream;
  */
 
 var JPEGStream = module.exports = function JPEGStream(canvas, options, sync) {
-  var self = this
-    , method = sync
+  if (!(this instanceof JPEGStream)) {
+    throw new TypeError("Class constructors cannot be invoked without 'new'");
+  }
+
+  Readable.call(this);
+
+  var self = this;
+  var method = sync
       ? 'streamJPEGSync'
       : 'streamJPEG';
   this.options = options;
   this.sync = sync;
   this.canvas = canvas;
-  this.readable = true;
+
   // TODO: implement async
   if ('streamJPEG' == method) method = 'streamJPEGSync';
-  process.nextTick(function(){
-    canvas[method](options.bufsize, options.quality, options.progressive, function(err, chunk){
-      if (err) {
-        self.emit('error', err);
-        self.readable = false;
-      } else if (chunk) {
-        self.emit('data', chunk);
-      } else {
-        self.emit('end');
-        self.readable = false;
-      }
-    });
-  });
-};
 
-/**
- * Inherit from `EventEmitter`.
- */
+  // For now we're not controlling the c++ code's data emission, so we only
+  // call canvas.streamPNGSync once and let it emit data at will.
+  var hasStarted = false;
+
+  self._read = function () {
+    if (!hasStarted) {
+      hasStarted = true;
+      process.nextTick(function(){
+        canvas[method](options.bufsize, options.quality, options.progressive, function(err, chunk){
+          if (err) {
+            self.emit('error', err);
+          } else if (chunk) {
+            self.push(chunk);
+          } else {
+            self.push(null);
+          }
+        });
+      });
+    }
+  };
```

Comment:
From earlier, we must use `process.nextTick`.

Sorry for not being more clear, or for still not understanding what you're saying, but this is not a hot path because it has the surrounding `if` block that causes evaluation of the function expression exactly once, thus there is nothing to reuse. I think then the only difference is style, but please let me know if I'm missing something.

As far as size of npm/inherits vs `util.inherits`, this is a node module only.